src/Features/Core/Portable/Diagnostics/EngineV2/CompilerDiagnosticExecutor.cs

```diff
@@ -19,41 +19,45 @@ internal static class CompilerDiagnosticExecutor
-            // REVIEW: we drop all those allocations for the diagnostics returned. can we avoid this?
-            await analyzerDriver.GetAnalyzerDiagnosticsAsync(cancellationToken).ConfigureAwait(false);
-
-            // this is wierd, but now we iterate through each analyzer for each tree to get cached result.
-            // REVIEW: no better way to do this?
-            var noSpanFilter = default(TextSpan?);
+            // PERF: Run all analyzers at once using the new GetAnalysisResultAsync API.
+            var analysisResult = await analyzerDriver.GetAnalysisResultAsync(cancellationToken).ConfigureAwait(false);
+
             var analyzers = analyzerDriver.Analyzers;
             var compilation = analyzerDriver.Compilation;
 
             var builder = ImmutableDictionary.CreateBuilder<DiagnosticAnalyzer, AnalysisResult>();
+
             foreach (var analyzer in analyzers)
             {
                 var result = new Builder(project, version);
 
-                // REVIEW: more unnecessary allocations just to get diagnostics per analyzer
-                var oneAnalyzers = ImmutableArray.Create(analyzer);
-
                 foreach (var tree in compilation.SyntaxTrees)
                 {
-                    var model = compilation.GetSemanticModel(tree);
-
-                    var syntax = await analyzerDriver.GetAnalyzerSyntaxDiagnosticsAsync(tree, oneAnalyzers, cancellationToken).ConfigureAwait(false);
-                    Contract.Requires(syntax.Count() == CompilationWithAnalyzers.GetEffectiveDiagnostics(syntax, analyzerDriver.Compilation).Count());
-                    result.AddSyntaxDiagnostics(tree, syntax);
+                    ImmutableDictionary<DiagnosticAnalyzer, ImmutableArray<Diagnostic>> syntaxDiagnosticsByAnalyzer;
+                    ImmutableArray<Diagnostic> syntaxDiagnostics;
+                    if (analysisResult.SyntaxDiagnostics.TryGetValue(tree, out syntaxDiagnosticsByAnalyzer) &&
+                        syntaxDiagnosticsByAnalyzer.TryGetValue(analyzer, out syntaxDiagnostics))
+                    {
+                        Contract.Requires(syntaxDiagnostics.Length == CompilationWithAnalyzers.GetEffectiveDiagnostics(syntaxDiagnostics, analyzerDriver.Compilation).Count());
+                        result.AddSyntaxDiagnostics(tree, syntaxDiagnostics);
+                    }
 
-                    var semantic = await analyzerDriver.GetAnalyzerSemanticDiagnosticsAsync(model, noSpanFilter, oneAnalyzers, cancellationToken).ConfigureAwait(false);
-                    Contract.Requires(semantic.Count() == CompilationWithAnalyzers.GetEffectiveDiagnostics(semantic, analyzerDriver.Compilation).Count());
-                    result.AddSemanticDiagnostics(tree, semantic);
+                    ImmutableDictionary<DiagnosticAnalyzer, ImmutableArray<Diagnostic>> semanticDiagnosticsByAnalyzer;
```

Comment:
syntaxDiagnosticsByAnalyzer/syntaxDiagnostics and corresponding semantic one seems could be shared.